### PR TITLE
[SPARK-33545][CORE] Support Fallback Storage during Worker decommission

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
### What changes were proposed in this pull request?

This PR aims to support storage migration to the fallback storage like cloud storage (`S3`) during worker decommission for the corner cases where the exceptions occur or there is no live peer left.

Although this PR focuses on cloud storage like `S3` which has a TTL feature in order to simplify Spark's logic, we can use alternative fallback storages like HDFS/NFS(EFS) if the user provides a clean-up mechanism.

### Why are the changes needed?

Currently, storage migration is not possible when there is no available executor. For example, when there is one executor, the executor cannot perform storage migration because it has no peer.

### Does this PR introduce _any_ user-facing change?

Yes. This is a new feature.

### How was this patch tested?

Pass the CIs with newly added test cases.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Could you review this, @holdenk , @viirya , @mridulm ?

File: core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        val name = ShuffleDataBlockId(shuffleId, mapId, NOOP_REDUCE_ID).name
+        val dataFile = new Path(fallbackPath, s"$shuffleId/$name")
+        val f = fallbackFileSystem.open(dataFile)
+        val size = nextOffset - 1 - offset
+        logDebug(s"To byte array $size")
+        val array = new Array[Byte](size.toInt)
+        f.seek(offset)
+        f.read(array)
+        val startTimeNs = System.nanoTime()
+        logDebug(s"Took ${(System.nanoTime() - startTimeNs) / (1000 * 1000)}ms")
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala:164]: Do you want to put `f.read` between `startTimeNs` and `logDebug`?

Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala:164]: Oops. Thanks.

**Comment by dongjoon-hyun** [comment]
It's a little weird because it passed locally.
```
$ build/sbt "core/testOnly *.FallbackStorageSuite"
...
[info] FallbackStorageSuite:
[info] - fallback storage APIs - copy/exists (957 milliseconds)
[info] - migrate shuffle data to fallback storage (224 milliseconds)
[info] - Upload from all decommissioned executors (5 seconds, 232 milliseconds)
[info] - Upload multi stages (3 seconds, 67 milliseconds)
[info] - Newly added executors should access old data from remote storage (8 seconds, 415 milliseconds)
[info] ScalaTest
[info] Run completed in 18 seconds, 975 milliseconds.
[info] Total number of tests run: 5
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 5, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
```

**Comment by Tagar** [comment]
@dongjoon-hyun is this only for shuffled data? I was wondering if it would also be possible to cover `MEMORY_AND_DISK` for cached dataframes? Thanks!

**Comment by dongjoon-hyun** [comment]
Thank you for review, @Tagar . Yes, they are separate options:
- `spark.storage.decommission.shuffleBlocks.enabled`
- `spark.storage.decommission.rddBlocks.enabled`

This PR is still under review. If this is accepted, I believe we can move on to that.

**Comment by dongjoon-hyun** [comment]
Could you review this PR once more please, @viirya ?

**Comment by holdenk** [comment]
Hi @dongjoon-hyun, I'm taking this week away from open source to take my puppy to go see snow for the first time. I'll do a review on Monday. Thanks for understanding :)

**Comment by dongjoon-hyun** [comment]
Thank you, @holdenk . Sorry for pinging you on the holiday season.

**Comment by dongjoon-hyun** [comment]
BTW, cc @dbtsai , too.

File: core/src/main/scala/org/apache/spark/storage/BlockManagerDecommissioner.scala [hunk 2]
```diff
@@ -114,6 +115,8 @@ private[storage] class BlockManagerDecommissioner(
                       // driver a no longer referenced RDD with shuffle files.
                       if (bm.migratableResolver.getMigrationBlocks(shuffleBlockInfo).isEmpty) {
                         logWarning(s"Skipping block ${shuffleBlockInfo}, block deleted.")
+                      } else if (fallbackStorage.isDefined) {
+                        fallbackStorage.foreach(_.copy(shuffleBlockInfo, bm))
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/storage/BlockManagerDecommissioner.scala:119]: When is the best time to move block to fallback storage during decommission? For current change, it just chooses fallback storage once encountering failure at first try.

Should we let retrying do it work until reaching `maxReplicationFailuresForDecommission` and then move to fallback storage?

Comment (viirya) [reply core/src/main/scala/org/apache/spark/storage/BlockManagerDecommissioner.scala:119]: Oh I see. You only put fallback storage into peers when there is no other peer executors.

File: core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    val mapId = shuffleBlockInfo.mapId
+
+    bm.migratableResolver match {
+      case r: IndexShuffleBlockResolver =>
+        val indexFile = r.getIndexFile(shuffleId, mapId)
+
+        if (indexFile.exists()) {
+          fallbackFileSystem.copyFromLocalFile(
+            new Path(indexFile.getAbsolutePath),
+            new Path(fallbackPath, s"$shuffleId/${indexFile.getName}"))
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala:62]: Is one `fallbackPath` possible shared between multiple Spark applications? If so will this path conflict between apps?

File: core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+          val dataFile = r.getDataFile(shuffleId, mapId)
+          if (dataFile.exists()) {
+            fallbackFileSystem.copyFromLocalFile(
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala:66]: The API `copyFromLocalFile` doesn't mention, but when the remote file exists, does it overwrite or throw exception?

File: core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+            val dataBlockId = ShuffleDataBlockId(shuffleId, mapId, reduceId)
+            FallbackStorage.reportBlockStatus(bm, dataBlockId, dataFile.length)
+          }
+        }
+      case r =>
+        logWarning(s"Unsupported Resolver: ${r.getClass.getName}")
+    }
+  }
+
+  def exists(shuffleId: Int, filename: String): Boolean = {
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala:85]: Is this only for test?

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 6]
```diff
@@ -627,7 +627,16 @@ private[spark] class BlockManager(
...
-      shuffleManager.shuffleBlockResolver.getBlockData(blockId)
+      try {
+        shuffleManager.shuffleBlockResolver.getBlockData(blockId)
+      } catch {
+        case e: IOException =>
+          if (conf.get(config.STORAGE_DECOMMISSION_FALLBACK_STORAGE_PATH).isDefined) {
+            FallbackStorage.read(conf, blockId)
+          } else {
+            throw e
+          }
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala:638]: Can you explain the reasoning here? So once any executor cannot get local shuffle block, it will always try to read from fallback storage?

File: core/src/main/scala/org/apache/spark/storage/BlockManagerDecommissioner.scala (hunk 2 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/storage/BlockManagerDecommissioner.scala:119]: Yes. It does.

File: core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala (hunk 3 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala:62]: No, `fallbackPath` should be unique in this PR and it will be injected at the job submission time.

File: core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala (hunk 4 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala:66]: Yes, it does and comply with S3 behavior.

File: core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala (hunk 5 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala:85]: It's exposed as a general API for the future although it's currently only verified at the UT.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala (hunk 6 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/storage/BlockManager.scala:638]: Yes. If `STORAGE_DECOMMISSION_FALLBACK_STORAGE_PATH` is defined, it tried to fallback. Currently, it happens when `FALLBACK_BLOCK_MANAGER_ID` has the blocks.

**Comment by dongjoon-hyun** [comment]
Thank you so much for your review, @viirya !

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala (hunk 6 above)
Comment (viirya) [reply core/src/main/scala/org/apache/spark/storage/BlockManager.scala:638]: Is there any check if `FALLBACK_BLOCK_MANAGER_ID` has the blocks before calling `FallbackStorage.read`? Looks like if `getBlockData` throws `IOException`, the block manager will just go to read from fallback storage.

Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/storage/BlockManager.scala:638]: It's already `IOException` and `FallbackStorage.read` throws `IOException` for non-exist files and it's fine and legitimate in the `WorkerDecomission` context, @viirya . Please note the following.
- The whole Worker decommission (including `storage decommission`) is designed as a best-effort approach.
- It's because the main use case is K8s graceful shutdown with the default period (`30s`). We can increase it, but we cannot set it infinite value technically. It means executor hangs in case of disk full situation.
- What we are aiming is to rescue data as much as possible, but `100%` is not guaranteed always.
- Lastly, data block selection(shuffle or rdd) was a random-order from the beginning. It became worse when there are multiple shuffle across multiple executors.

Due to the above reasons, we introduced `SPARK-33387 Support ordered shuffle block migration` to rescue as complete as possible. However, we can still easily imagine that multiple shuffles coexist on an executor in a skewed manner. While the other executor succeeds to complete the migration, the skewed executor may fail to complete the migration in the given grace period.

File: core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+        if (indexFile.exists()) {
+          fallbackFileSystem.copyFromLocalFile(
+            new Path(indexFile.getAbsolutePath),
+            new Path(fallbackPath, s"$shuffleId/${indexFile.getName}"))
+
+          val dataFile = r.getDataFile(shuffleId, mapId)
+          if (dataFile.exists()) {
+            fallbackFileSystem.copyFromLocalFile(
+              new Path(dataFile.getAbsolutePath),
+              new Path(fallbackPath, s"$shuffleId/${dataFile.getName}"))
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala:68]: Should we add the application id to the path? For example, a user may just run the same codes multiple times. It would be more convenient if the user doesn't need to use a different configuration value for each run.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 8]
```diff
@@ -471,6 +471,15 @@ package object config {
         "cache block replication should be positive.")
       .createWithDefaultString("30s")
 
+  private[spark] val STORAGE_DECOMMISSION_FALLBACK_STORAGE_PATH =
+    ConfigBuilder("spark.storage.decommission.fallbackStorage.path")
+      .doc("The location for fallback storage during block manager decommissioning." +
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:476]: nit: it's better to remind the user to enable TTL in the doc since Spark doesn't clean up the fallback storage.

File: core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala (hunk 7 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala:68]: Thank you for review, @zsxwing . @viirya also asked the similar [comments](https://github.com/apache/spark/pull/30492#discussion_r531170833). And, right, currently, the assumption is the runner (or the users) injects the unique path for each submissions for all cases. Let me think about that requirement (@zsxwing and @viirya )

File: core/src/main/scala/org/apache/spark/internal/config/package.scala (hunk 8 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala:476]: Sure, @zsxwing !

**Comment by dongjoon-hyun** [comment]
Hi, @zsxwing and @viirya . I addressed your comments. Could you review once more please if you have a chance?

**Comment by zsxwing** [comment]
Thanks for adding the application id. I don't have time to look at details recently. Will defer that to @viirya

**Comment by dongjoon-hyun** [comment]
Got it, @zsxwing .

**Comment by dongjoon-hyun** [comment]
Hi, @holdenk and @viirya and @dbtsai . Could you review this, please?

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 9]
```diff
@@ -627,7 +627,16 @@ private[spark] class BlockManager(
   override def getLocalBlockData(blockId: BlockId): ManagedBuffer = {
     if (blockId.isShuffle) {
       logDebug(s"Getting local shuffle block ${blockId}")
-      shuffleManager.shuffleBlockResolver.getBlockData(blockId)
+      try {
+        shuffleManager.shuffleBlockResolver.getBlockData(blockId)
+      } catch {
+        case e: IOException =>
```
Comment (holdenk) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala:633]: Nit/ question: Could we move the if up as part of the case and avoid the need for explicit rethrow?

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala (hunk 6 above)
Comment (holdenk) [reply core/src/main/scala/org/apache/spark/storage/BlockManager.scala:638]: I think this would probably fail quickly if the external storage doesn’t have the block right? So no need to do a separate check to see if the block is present, the failure message is good enough.

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala [hunk 10]
```diff
@@ -295,8 +295,9 @@ final class ShuffleBlockFetcherIterator(
     var hostLocalBlockBytes = 0L
     var remoteBlockBytes = 0L
 
+    val fallback = FallbackStorage.FALLBACK_BLOCK_MANAGER_ID.executorId
     for ((address, blockInfos) <- blocksByAddress) {
-      if (address.executorId == blockManager.blockManagerId.executorId) {
+      if (Seq(blockManager.blockManagerId.executorId, fallback).contains(address.executorId)) {
```
Comment (holdenk) [inline core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:300]: I’m fuzzy on why this change is needed, would the fallback ever be equal to the execid?

**Review by holdenk — COMMENTED** [review]
Thanks for waiting on the review. I’ve got a few minor nits/ questions but overall this looks good to me. I’m excited for us to support this as part of dynamic scale down in Spark 3.1 :)

File: core/pom.xml [hunk 11]
```diff
@@ -461,6 +461,47 @@
       <scope>test</scope>
     </dependency>
 
+    <dependency>
+      <groupId>org.apache.hadoop</groupId>
+      <artifactId>hadoop-aws</artifactId>
+      <version>${hadoop.version}</version>
+      <scope>test</scope>
```
[the remark this answers is not in the document]
Comment (dongjoon-hyun) [reply core/pom.xml:468]: Thank you for review. It's the same reason for Spark history location (`spark-events`). If the user want to use cloud-storage, they will provide it at runtime via `-Phadoop-cloud` or `--package`. If the user provides a TTL feature, I believe they can use any locations like HDFS/EFS.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 12]
```diff
@@ -471,6 +471,16 @@ package object config {
         "cache block replication should be positive.")
       .createWithDefaultString("30s")
 
+  private[spark] val STORAGE_DECOMMISSION_FALLBACK_STORAGE_PATH =
+    ConfigBuilder("spark.storage.decommission.fallbackStorage.path")
+      .doc("The location for fallback storage during block manager decommissioning. " +
+        "For example, `s3a://spark-storage/`. In case of empty, fallback storage is disabled. " +
+        "The storage should be managed by TTL because Spark will not clean up it.")
```
[the remark this answers is not in the document]
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala:478]: Oops. Got it.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala (hunk 9 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/storage/BlockManager.scala:633]: It's because we should access the normal access path by default. We are able to use `Fallback` path only if the normal access path fail because `conf.get(config.STORAGE_DECOMMISSION_FALLBACK_STORAGE_PATH).isDefined` doesn't mean the fallback storage has the data.

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala (hunk 10 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:300]: Yes. This is required for simplicity to make Spark consider the external blocks as a local block.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala (hunk 9 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/storage/BlockManager.scala:633]: Please let me know if I misunderstand your comment.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala (hunk 6 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/storage/BlockManager.scala:638]: Yes, it does, @holdenk .

Comment (viirya) [reply core/src/main/scala/org/apache/spark/storage/BlockManager.scala:638]: Make sense. Thanks @holdenk

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 13]
```diff
@@ -471,6 +471,16 @@ package object config {
         "cache block replication should be positive.")
       .createWithDefaultString("30s")
 
+  private[spark] val STORAGE_DECOMMISSION_FALLBACK_STORAGE_PATH =
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/internal/config/package.scala:474]: Because we only register fallback storage when SparkContext initialization, I think it doesn't make sense to change this config after creating SparkContext, so this sounds like a static SQL config?

Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala:474]: This is `CORE` module which can work without SQL module. So, we cannot put it to Static SQL conf, @viirya .

Comment (viirya) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala:474]: Oh I see. Seems we don't have static config for core.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala (hunk 9 above)
Comment (holdenk) [reply core/src/main/scala/org/apache/spark/storage/BlockManager.scala:633]: Yeah so I mean we still have a try/catch just the case statement has the if as part of it (eg “ Using if expressions in case statements”)

File: core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala (hunk 10 above)
Comment (holdenk) [reply core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala:300]: Sounds reasonable

**Review by holdenk — APPROVED** [review]
LGTM

File: core/src/main/scala/org/apache/spark/internal/config/package.scala (hunk 13 above)
Comment (viirya) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala:474]: Could you add a simple note to the doc? E.g., "Fallback storage is registered during SparkContext initialization, so this must be enabled before creating SparkContext."

**Review by viirya — COMMENTED** [review]
Looks okay to me.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala (hunk 13 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala:474]: Ur, but it will be the same for all decommission conf like `spark.storage.decommission.enabled` and the other general conf like `spark.driver.host` and `spark.driver.port`.

**Comment by dongjoon-hyun** [comment]
Thank you so much, @holdenk and @viirya .
Merged to master for Apache Spark 3.1.0.

File: core/src/main/scala/org/apache/spark/internal/config/package.scala (hunk 13 above)
Comment (viirya) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala:474]: Oh okay.

File: core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+    } else {
+      None
+    }
+  }
+
+  /** Register the fallback block manager and its RPC endpoint. */
+  def registerBlockManagerIfNeeded(master: BlockManagerMaster, conf: SparkConf): Unit = {
+    if (conf.get(STORAGE_DECOMMISSION_FALLBACK_STORAGE_PATH).isDefined) {
+      master.registerBlockManager(
+        FALLBACK_BLOCK_MANAGER_ID, Array.empty[String], 0, 0, new NoopRpcEndpointRef(conf))
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/storage/FallbackStorage.scala:118]: Just a question. Do we need to do the registration? The fallback block manager is just a place holder.
